Client/Application/AppDelegate.swift

```diff
@@ -96,6 +108,36 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
                     }
                 }
             }
+        } else if let components = NSURLComponents(URL: url, resolvingAgainstBaseURL: false) where components.scheme == "firefox" && components.host == "x-callback-url" {
+            if let query = components.query, items = components.queryItems as? [NSURLQueryItem] {
+                if items[0].name == "x-source" && items[0].value != nil && items[1].name == "url" && items[1].value != nil {
+                    if let newURL = NSURL(string: items[1].value!) {
+                        let tab = self.tabManager.addTab(request: NSURLRequest(URL: newURL))
+                        self.tabManager.selectTab(tab)
+                        if let backToAppURL = NSURL(string: items[0].value!) {
+                            if let appName = items[2].value where items[2].name == "x-source-name" {
+                                if UIApplication.sharedApplication().canOpenURL(backToAppURL) {
```

Comment:
this is checking to see if the app can be called back from firefox; I already open the tab before the backToAppURL check